### ISSUE TITLE
Mobile navigation menus appear like dialogs

### ISSUE THREAD

**Issue by adriannelee** [cds-snc/notification-planning#582]
Issue group: Header
Visibility: 3
Evaluation: Partially Supports
Success Criteria: 1.3.1: Info and Relationships
Needs design: Interaction

### Description
The mobile navigation menus are presented like dialogs, although missing the necessary aria roles. However, it may be more understandable to treat the mobile menu as a disclosure navigation pattern. Some related changes would be to preserve the visibility of the original button as a means to close the menu, and tabbing through the menu would also close it. It may be worthwhile to test out the current and suggested mobile menu implementations with users testing.

https://www.w3.org/TR/wai-aria-practices/examples/disclosure/disclosure-navigation.html

### Acceptance Criteria 
- [x] overlay menus on mobile work for people using screen readers 
- [x] automated a11y tests
- [x] manually tested with screen readers and dictation software 
- [x] Add missing aria roles
- [x] when you open the menu, the focus should stay on the menu or the first item of the menu instead of staying on the button you select when you go to the menu

![](https://dl.airtable.com/.attachments/b0b68af336ef37078a178b9969db06aa/c7e6a13b/Capturedecranle2022-03-15a09_48_24.png)

**Comment by amazingphilippe** [comment]
- Some decision on what solution: proper aria or another pattern
- Smush the menus back together

**Comment by andrewleith** [comment]
Changing this to a disclosure navigation pattern requires quite a bit of re-work in the code as well as visually.  It's possible it may make the menus easier to use, but I'm not sure it is worth the effort.

I think it would be best to just fix the issues that prevent this from being implemented correctly as a dialog instead, i.e. marking it up correctly using the proper roles and keyboard shortcuts.

@adriannelee, @amazingphilippe please let me know your thoughts.

**Comment by adriannelee** [comment]
@amazingphilippe do we know why the auditor recommended the disclosure pattern? instead of a dialogue?

**Comment by amazingphilippe** [comment]
Reading back at it, the report even suggested testing both with users. I think we should make a decision based on the effort we'd like to put in this. We've already put some effort in designing this to look like a disclosure pattern. 

Another way we could look at this is the effort of maintaining the main navigation. I know we have 2 navs in the code, could this work be an opportunity to reduce to 1? and remove some content duplication in the code? In the long run, this would be less effort.

**Comment by amazingphilippe** [comment]
From a visual point of view, maybe the auditor is suggesting that a disclosure pattern is less disruptive to the natural flow of the page? He mentions the menu button remaining in place on toggle. This is how I would interpret their suggestion. On another note, I think keeping the FIP could help orient the users back and forth between the expanded and collapsed states.

**Comment by yaelberger-commits** [comment]
Phil and Andrew meeting this morning to figure out next steps on this one

**Comment by amazingphilippe** [comment]
Given the work going on for multiple ways, it makes sense to come up with a more robust js module for disclosure patterns based on a standard button and nav elements.

Some good example sources:

The codepen in this article: https://www.w3.org/WAI/ARIA/apg/example-index/disclosure/disclosure-navigation.html
This library. We already do the same, but more custom: https://gijsroge.github.io/priority-nav.js/#

**Comment by yaelberger-commits** [comment]
Phil and Andrew met Thursday on this and have a plan to move forward.
Looking for one set of menus that are responsible, disclosure pattern at the same time

**Comment by jimleroyer** [comment]
Yael, Andrew and Phil to sync! (maybe Jimmy?)

**Comment by yaelberger-commits** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Notify-Planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/582?planning-poker) with ZenHub @jzbahrai @sastels

**Comment by yaelberger-commits** [comment]
@andrewleith @jimleroyer @jzbahrai & @sastels please add your estimates for planning poker to this one

**Comment by amazingphilippe** [comment]
Reviewed and merged. Waiting for next release

**Comment by amazingphilippe** [comment]
Continuing this work with #761. We'll compare disclosure and overlay navigations

**Comment by yaelberger-commits** [comment]
@amazingphilippe and @andrewleith Can you please update the Acceptance Criteria checklist as you go through QA?
Thanks

**Comment by amazingphilippe** [comment]
Found an error with the dialog. We place a label on a div and that's not allowed! I can't find direct quote on this, but it seems aria-labelledby is preferred for dialogs. 

> Error: aria-label is used on an element which does not allow this attribute

**Comment by andrewleith** [comment]
> Found an error with the dialog. We place a label on a div and that's not allowed! I can't find direct quote on this, but it seems aria-labelledby is preferred for dialogs. (compare aria-label and aria labelledby on [this page](https://www.w3.org/WAI/ARIA/apg/example-index/))
> 
> > Error: aria-label is used on an element which does not allow this attribute
> 
> Found with ARC toolkit

Yeah looks like you can't use `aria-label` on non-interactive content, but you can on particular roles (such as `navigation` [and possibly `dialog` since it isn't complaining about `<div role="dialog" aria-label="Main menu">`]).  @amazingphilippe where are you seeing the issue?  I don't see any issues on the main menu.  Is it this:
```
<div aria-label="Your account and language" class="flex justify-end self-center">
```
If so, we can probably just omit that entirely and move it to the interactive element just below:
```
<button aria-expanded="false" aria-haspopup="true" id="account-icon-mobile" class="p-1 h-14 border-2 border-solid border-gray rounded-lg" data-module="menu-overlay" data-menu-overlay-content="account-menu-mobile" data-menu-overlay-close="account-close-mobile">
    <img aria-hidden="false" aria-label="Account menu" src="https://staging.notification.cdssandbox.xyz/static/images/account-icon.svg?c09b09a662f3ee41af2ce9bad05ea1f4" alt="Account menu" class="w-10 h-10">
  </button>
```

Looking closer at that snippet, is there a reason we used `aria-label` on the `img` tag?  I would think it should be on the interactive element instead (the button).

**Comment by andrewleith** [comment]
> Axe dev tools found 2 more issues:
> 
> 1. [Screen readers can read content outside the modal dialog.](https://axe.deque.com/issues/e08e3353-3090-4af5-a648-bf17e1c78a4c?lang=en)
> 2. [When the modal dialog or similar element is closed, keyboard focus is not returned to the triggering element.](https://axe.deque.com/issues/b00961c8-300b-4f1a-a94a-cce3cdc0b06a?lang=en)

Local axe dev tools didn't report either of these issues.  Maybe because I'm using the free version?
Also just want to note that these are both pre-existing issues.

**Comment by andrewleith** [comment]
@amazingphilippe yeah I think so too. Moving the `aria-label` from the `div` down to the `button` and removing in entirely from the `img` should satisfy all of this.  I'm not sure why the tools didn't complain about the `aria-label` on the `img` tag, since it isn't interactive either.

@yaelberger-commits would you like me to tackle the remaining issues in this sprint? The menu ones are easy but the focus ones are likely a little tougher:

Menu issues
- [ ] Move `aria-label` from div onto following `<button>` element 
- [ ] Remove the`aria-label` from the img tag

Modal issues 
- [ ] Return focus to menu link when menu closes
- [ ] Prevent screen readers from reading content outside of modal

**Comment by adriannelee** [comment]
Might split off menu issues and modal issues into separate cards

**Comment by yaelberger-commits** [comment]
I think splitting them off with the modal/focus issues in a separate card sounds like the path forward. 
We can continue tackling the aria issues this sprint and see about the Modal ones for next perhaps

**Comment by andrewleith** [comment]
Menu issues addressed in cds-snc/notification-admin#1302

**Comment by amazingphilippe** [comment]
Created a new issue for the remaining focus management work. Tested one last time and looks good!
